### DISCUSSION TITLE
BLE.config 'rxbuf'  in v1.20

### DISCUSSION THREAD

**Discussion by Pawelinode** [micropython/micropython#11865] [ESP32]
In Micropython 1.20 I can't set 'rxbuf'. Has something changed?

```
import bluetooth
ble=bluetooth.BLE()
if ble.active()==False:
    ble.active(True)
ble.config(rxbuf=256)

```
Traceback (most recent call last):
  File "<stdin>", line 5, in <module>
ValueError: unknown config param

**Comment by jimmo** [comment]
@Pawelinode `rxbuf` was only necessary to set the ring buffer size (allowing a higher rate of incoming messages). In v1.20 we improved the BLE implementation on ESP32 to no longer require the ring buffer, so this parameter is no longer needed.

**Reply by puppet13th** [reply]
i'm using esp32 as uart peripheral and android as client using serial bluetooth terminal.
on connection esp32 mtu exchanged to 256.
esp32 can send more than 20 bytes of data but cannot received more than 20 bytes of data.
![image](https://github.com/micropython/micropython/assets/48564964/d697f838-69da-4b34-918d-8a5d61a66cf4)

code output:
```
Connection from Device(ADDR_RANDOM, 61:87:75:e0:a3:5c, CONNECTED)
mtu exchanged. 256
Device(ADDR_RANDOM, 61:87:75:e0:a3:5c, CONNECTED) 12345678901234567890 <-- only 20 bytes data received from android

```

look like default buffer for ble characteristic is 20 bytes. to increase the buffer just need to write the ble characteristic.
using aioble:
`uart_rx.write( bytes( 32 ) )`

now my esp32 uart peripheral can received up to 32 bytes of data

**Reply by nestorld** [reply]
Sorry @jimmo, could you please elaborate more on how I can receive more than 20 characters? I have two ESP32, one is acting as a peripheral and the other one is a central with micropython 1.20. With a read characteristic of the peripheral, I can send long messages, but what the central is receiving there is truncated to 20 characters. I don't see how to increase this limit without the rxbuf config param

**Reply by puppet13th** [reply]
> Sorry @jimmo, could you please elaborate more on how I can receive more than 20 characters? I have two ESP32, one is acting as a peripheral and the other one is a central with micropython 1.20. With a read characteristic of the peripheral, I can send long messages, but what the central is receiving there is truncated to 20 characters. I don't see how to increase this limit without the rxbuf config param

The default size of write is 20. If u remote to write more than 20 bytes. Just write to local char.

Example when using aioble module:
`char.write(bytes( 64))`
this should increase that char to received upto 64 bytes.

**Reply by nestorld** [reply]
But I can not write from the central to a read/notify characteristic of the peripheral

**Reply by puppet13th** [reply]
the peripheral device in question should write to it's char to break 20 bytes limit.

**Reply by jimmo** [reply]
Yep exactly what @puppet13th said. The server should do this (i.e. write to its own characteristic) just after registering the characteristic.
Then when the client sends a remote write, the buffer on the server will be large enough.

@nestorld In the latest version of `aioble` you can also use `BufferedCharacteristic(..., max_len=N)` in place of `Characteristic(...)` when registering. i.e. if your code previously had

```py
temp_char = aioble.Characteristic(temp_service, _ENV_SENSE_TEMP_UUID, read=True, notify=True)
```

you can now do

```py
temp_char = aioble.BufferedCharacteristic(temp_service, _ENV_SENSE_TEMP_UUID, read=True, notify=True, max_len=50)
```

**Comment by nestorld** [comment]
It is doing it, but the central is receiving 20 bytes

**Comment by nestorld** [comment]
Sorry Jim, the issue is in the opposite way, the client (central)  is
receiving 20 bytes. That is not the one registering the characteristic

**Reply by jimmo** [reply]
Then you need to set the MTU

**Reply by nestorld** [reply]
I'm already doing this when starting the central:
        self.ble.active(True)
        self.ble.irq(self.ble_irq)
        self.ble.config(mtu=512)

BTW, I'm using bluetooth.BLE

The peripheral is working well, I'm also already writing 512 bytes when registering its characteristics. If I test connecting to the peripheral from a testing tool (LightBlue), I can perfectly communicate with it and receive data with way more than 20 bytes in LightBlue (the central in this test). But when I replace LightBlue with the central I'm designing, the data is truncated to 20 bytes.
Then, this says to me that there is nothing to do with the peripheral, everything is ok with it, but I need to do something on the central  to let it handle more than 20 bytes, and that is what I'm not seeing how to do.

**Reply by nestorld** [reply]
Ok, the solution was here, related to the mtu as indicated by @jimmo. Just after connecting to the peripheral, I had to execute this from the central to negotiate the mtu:

     ble.gattc_exchange_mtu(blePeriferal.device.connHandle)

Previously I had set the mtu in both the peripheral and central to be 512 (ble.config(mtu=512)), but it is taken into account only after it is negotiated using the ble.gattc_exchange_mtu method from the central. With that, the minimum mtu between the peripheral and central will be used to communicate as indicated in the documentation:

" 'mtu': Get/set the MTU that will be used during a ATT MTU exchange. The resulting MTU will be the minimum of this and the remote device’s MTU. ATT MTU exchange will not happen automatically (unless the remote device initiates it), and must be manually initiated with [gattc_exchange_mtu](https://docs.micropython.org/en/latest/library/bluetooth.html#bluetooth.BLE.gattc_exchange_mtu). Use the _IRQ_MTU_EXCHANGED event to discover the MTU for a given connection."

https://docs.micropython.org/en/latest/library/bluetooth.html#bluetooth.BLE.gattc_exchange_mtu

     Thank you guys for helping on this!

**Comment by l00jj** [comment]
Thank all guys, But it still took me a long time to explore, and I have summarized it
esp32c3, API & bin v1.24.0

https://github.com/l00jj/MicroPython/wiki/bluetooth.BLE

The key point is to set MTU when connecting devices, and set the MTU config and gatts buffer size before proceeding
